File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

func SendAll(socket io.Writer, bytes []byte) error {
	total := 0

	for {
		n, err := socket.Write(bytes[total:])

		total += n

		if total == len(bytes) {
			return nil
		}

		if err != nil {
			return err
		}

	}

}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	total := 0
	for {
		n, err := socket.Read(buff[total:])

		total += n

		if total == size {
			return buff, nil
		}

		if err != nil {
			return buff[:total], err
		}

	}

}

File: services/server/src/safe_socket/safe_socket.py
import socket


def recv_all(sock: socket.socket, size):
    total = 0 
    buff = bytearray(size)
    while total < size:
        try:
            n = sock.recv(size - total)
        except socket.error:
            raise RuntimeError("Socket connection closed")

        buff[total:total+len(n)] = n

        if len(n) == 0:
            raise RuntimeError("Socket connection closed")

        total += len(n)
    return buff


def send_all(sock: socket.socket, bytes):
    total = 0 
    while total < len(bytes):
        try:
            n = sock.send(bytes[total:])
        except socket.error:
            raise RuntimeError("Socket connection closed")

        total += n
    return total
